Adding AWS S3 Example (#681)
nautobot/nautobot: examples/s3_static_files/README.md

TLDR:

```shell
$ echo django-storages >> $NAUTOBOT_ROOT/local_requirements.txt
$ pip3 install django-storages
```

## Configuration

The [`django-storages`](https://django-storages.readthedocs.io/en/stable/) library is quite powerful, please refer to their documentation for a more detailed explanation of the individual settings or for more information.

In `nautobot_config.py` define the following configuration:

```python
STORAGE_BACKEND = "storages.backends.s3boto3.S3Boto3Storage"

STORAGE_CONFIG = {
    "AWS_ACCESS_KEY_ID": "...",
    "AWS_SECRET_ACCESS_KEY": "...",
    "AWS_STORAGE_BUCKET_NAME": "my-bucket-name",
    "AWS_S3_REGION_NAME": "us-west-1",
    "AWS_DEFAULT_ACL": "public-read",
    "AWS_QUERYSTRING_AUTH": False,
    "AWS_LOCATION": "subfolder/name/"
}
STATICFILES_STORAGE = STORAGE_BACKEND
```

If `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` are not set, `boto3` [internally looks up IAM credentials](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html).  Using an [IAM Role for EC2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html?icmpid=docs_ec2_console) is highly recommended.

## Bucket Creation

> The AWS S3 bucket will be hosting Nautobot static files and needs some specific configuration to allow anonymous HTTP access.  The following is the necessary Terraform configuration to create the S3 bucket appropriately:

Reviewer: The mention of Terraform here appears to be non-sequitur? Is this specific to how we are using S3 with Nautobot?

Author: It's provided as an example of how it might be done.  Documenting the manual process of setting the correct bucket settings involves a lot of steps

Reviewer 2: Perhaps change the wording to something like "The following is an example of Terraform configuration to create the S3 bucket appropriately"?

# AWS S3 Static Files

This example shows how you can configure Nautobot to host static files on AWS S3.

## Installation

> Nautobot uses [`django-storages`](https://django-storages.readthedocs.io/en/stable/) to publish files to s3 see the [docs](https://nautobot.readthedocs.io/en/stable/installation/nautobot/#remote-file-storage) for more information.

Reviewer 2: ```suggestion
Nautobot can use [`django-storages`](https://django-storages.readthedocs.io/en/stable/) to publish files to S3 -- see the [docs](https://nautobot.readthedocs.io/en/stable/installation/nautobot/#remote-file-storage) for more information.
```

TLDR:

```shell
$ echo django-storages >> $NAUTOBOT_ROOT/local_requirements.txt
$ pip3 install django-storages
```

## Configuration

The [`django-storages`](https://django-storages.readthedocs.io/en/stable/) library is quite powerful, please refer to their documentation for a more detailed explanation of the individual settings or for more information.

In `nautobot_config.py` define the following configuration:

```python
STORAGE_BACKEND = "storages.backends.s3boto3.S3Boto3Storage"

STORAGE_CONFIG = {
    "AWS_ACCESS_KEY_ID": "...",
    "AWS_SECRET_ACCESS_KEY": "...",
    "AWS_STORAGE_BUCKET_NAME": "my-bucket-name",
    "AWS_S3_REGION_NAME": "us-west-1",
    "AWS_DEFAULT_ACL": "public-read",
    "AWS_QUERYSTRING_AUTH": False,
    "AWS_LOCATION": "subfolder/name/"

>     "USE_TZ": True,

Reviewer 3: Could we potentially just do:

```
    "USE_TZ": USE_TZ,
```

It looks like the nautobot_config.py imports in the settings. Not sure if the potential for someone to set the global `USE_TZ` to something different than the `USE_TZ` stored within `STORAGE_CONFIG`.

Author: I think I'd either rather set it explicitly to `True` here or wait until PR #673 is merged and just remove it here

Reviewer 3: That works. I'm not sure if #673 will make it into 1.1.0 or if we're going to get rid of the monkey patching completely in a later release.

}
STATICFILES_STORAGE = STORAGE_BACKEND
```

> If `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` are not set, boto3 internally looks up IAM credentials.  Using an [IAM Role for EC2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html?icmpid=docs_ec2_console) is highly recommended.

Reviewer 2: "internally looks up IAM credentials..." from where?

```suggestion
If `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` are not set, `boto3` internally looks up IAM credentials.  Using an [IAM Role for EC2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html?icmpid=docs_ec2_console) is highly recommended.
```

## Bucket Creation

> The AWS S3 bucket will be hosting Nautobot static files and needs some specific configuration to allow anonymous http access.  The following is the necessary terraform configuration to create the s3 bucket appropriately:

Reviewer 2: ```suggestion
The AWS S3 bucket will be hosting Nautobot static files and needs some specific configuration to allow anonymous HTTP access.  The following is the necessary Terraform configuration to create the S3 bucket appropriately:
```
